akka-distributed-data/src/main/scala/akka/cluster/ddata/DurableStore.scala

```diff
@@ -0,0 +1,245 @@
+ * `Load` message to it. It must then reply with 0 or more `LoadData` messages
+ * followed by one `LoadCompleted` message to the `sender` (the `Replicator`).
+ *
+ * When the `Replicator` needs to store a value it sends a `Store` message
+ * to the durable store actor, which must then reply with the `successMsg` or
+ * `failureMsg` to the `replyTo`.
+ */
+object DurableStore {
+  final case class Store(key: String, data: ReplicatedData, reply: Option[StoreReply])
+  final case class StoreReply(successMsg: Any, failureMsg: Any, replyTo: ActorRef)
+
+  case object Load
+  final case class LoadData(data: Map[String, ReplicatedData])
+  case object LoadCompleted
+}
+
+object MapDbDurableStore {
+  def props(config: Config): Props =
+    Props(new MapDbDurableStore(config))
+
+  private case object Commit extends DeadLetterSuppression
+}
+
+final class MapDbDurableStore(config: Config) extends Actor with ActorLogging {
+  import DurableStore._
+  import MapDbDurableStore.Commit
+
+  val serializer = SerializationExtension(context.system).serializerFor(classOf[DataEnvelope])
+  val commitInterval = config.getString("mapdb.commit-interval").toLowerCase match {
+    case "off" ⇒ Duration.Zero
+    case _     ⇒ config.getDuration("mapdb.commit-interval", MILLISECONDS).millis
+  }
+
+  val db = {
+    val file = config.getString("mapdb.file") match {
+      case path if path.endsWith(".ddata") ⇒
+        new File(path)
+      case dir ⇒
+        new File(dir, s"${context.system.name}-${self.path.parent.name}-${Cluster(context.system).selfAddress.port.get}.ddata")
+    }
+    file.getParentFile.mkdirs()
+
+    val builder = DBMaker.fileDB(file)
+      .fileMmapEnableIfSupported()
+      .fileMmapPreclearDisable()
+      .allocateStartSize(config.getBytes("mapdb.allocate-start-size"))
+      .allocateIncrement(config.getBytes("mapdb.allocate-increment"))
+      .concurrencyDisable() // only access from this actor
+      .closeOnJvmShutdown()
+
+    if (config.getBoolean("mapdb.write-ahead-log"))
+      builder.transactionEnable() // WAL for crash safety
+
+    builder.make()
+  }
+  val store: HTreeMap[String, ReplicatedData] =
+    db.hashMap("store")
+      .keySerializer(Serializer.STRING)
+      .valueSerializer(new MapdbReplicatedDataSerializer(context.system))
+      .createOrOpen()
+
+  val uncommitted = new java.util.HashMap[String, ReplicatedData]
+
+  override def postStop(): Unit = {
+    super.postStop()
+    commit()
+    db.close()
+  }
+
+  def receive = init
+
+  def init: Receive = {
+    case Load ⇒
+      try {
+        val loadData = LoadData(store.entrySet.asScala.map { entry ⇒
+          entry.getKey → entry.getValue
+        }(breakOut))
+        if (loadData.data.nonEmpty)
+          sender() ! loadData
+        sender() ! LoadCompleted
+      } catch {
+        case NonFatal(e) ⇒
+          log.error(e, "failed to load data")
+          sender() ! LoadCompleted
```

Comment:
yes, I'm not sure what is best to do if load fails. Stop everything? Perhaps that is better, and they have to remove the file to startup. I don't think we can stop the actor system because of this, so we can only stop the Replicator and an application would not really notice unless they watch it (except that they will get no replies). Perhaps that is the best we can do anyway.